### ISSUE TITLE
Support `call_fun2` Beam instruction

### ISSUE THREAD

**Issue by dumbbell** [rabbitmq/khepri#19]
While testing Khepri with Erlang master branch (the future Erlang 25), @mkuratczyk saw that Khepri's function extraction and transaction support doesn't support the new `call_fun2` instruction.

Beside while-listing it, we need to study it and determine if Khepri must follow the call and copy/extract more code.

**Comment by the-mikedavis** [comment]
This test case creates a `call_fun2`:

```erl
apply_fun_to_args(Fun, Arg1, Arg2) ->
    Fun(Arg1, Arg2).

allowed_higher_order_external_call_test() ->
    ?assertStandaloneFun(
        begin
            Fun = fun erlang:min/2,
            1 = apply_fun_to_args(Fun, 1, 2)
        end).
```

With a practical example being something like [this from `khepri.erl`](https://github.com/rabbitmq/khepri/blob/62197b4ac85351bef9501d17bb4b0a28c8b041f1/src/khepri.erl#L419).

In asm this looks like:

```erl
{function, apply_fun_to_args, 3, 240}.
  {label,239}.
    {line,[{location,"test/tx_funs.erl",493}]}.
    {func_info,{atom,tx_funs},{atom,apply_fun_to_args},3}.
  {label,240}.
    {'%',{var_info,{x,0},[{type,{t_fun,2,any,any}}]}}.
    {'%',{var_info,{x,1},[{type,{t_integer,{1,1}}}]}}.
    {'%',{var_info,{x,2},[{type,{t_integer,{2,2}}}]}}.
    {allocate,0,1}.
    {move,{integer,2},{x,1}}.
    {move,{x,0},{x,2}}.
    {move,{integer,1},{x,0}}.
    {line,[{location,"test/tx_funs.erl",494}]}.
    {call_fun2,{atom,safe},2,{tr,{x,2},{t_fun,2,any,any}}}.
    {deallocate,0}.
    return.
```

But the disasm looks a bit mangled:

```erl
{call_fun2,{atom,safe},2,{tr,{x,2},{{t_fun,any,any,any},0,18446744073709551615}}
```

I think we should be able to discard the wrapping `{Type, 0, 18446744073709551615}` and just use `Type` but I'm not sure yet where it comes from.

This is only one of the cases for the `Tag` (element 2) though. `khepri_machine` has some examples of `{atom,unsafe}` that should probably be tested. There's a [third case](https://github.com/erlang/otp/blob/ce4ee6e960ece4193fd3031880f8cfc6b50ee69f/lib/compiler/src/genop.tab#L661) that I think should look like `{f, 123}` and will need to have its label be rewritten. I haven't found a way to produce that instruction yet though: passing in a captured local function just results in a `call` or `call_only`.

**Comment by dumbbell** [comment]
Thank you for your research in this instruction!

We must ensure in a testcase that `Fun` was followed, extracted and verified by `khepri_fun`. Then it might be ok to allow `call_fun2`. Beside removing the extra arguments to this instruction, I wonder if we need to patch the `{t_fun, ...}` tuple: I see a `2` from the compiler vs `any` from `beam_disasm` as the second element in that tuple. Is it the arity of the function?

As a starting point, we can only allow `{call_fun2, {atom, safe}, ...}` and deny other variants.

> I think we should be able to discard the wrapping {Type, 0, 18446744073709551615}

That happens a lot with what `beam_disasm` produces unfortunately... In the Beam encode instruction, there are fields used to decode other fields (like the number of elements in an array, an offset or size, ...) and `beam_disasm` add them to the reconstructed instruction even though they don't make any sense at this point and are even incorrect from a compiler point of view.

**Comment by michaelklishin** [comment]
We should report this as a `beam_disasm` bug then?

**Comment by the-mikedavis** [comment]
> We must ensure in a testcase that `Fun` was followed, extracted and verified by `khepri_fun`

It looks like we'll need some extra checks for `move/2` to cover captures in order to track down `Fun`:

```erl
-module(callfunsmuggle).

-export([send_msg/3]).

do_apply(Fun, Arg1, Arg2) ->
    Fun(Arg1, Arg2).

send_msg(Pid, Msg) ->
    Fun = fun erlang:send/2,
    do_apply(Fun, Pid, Msg).

%% becomes the following (with extra instructions removed):
{function, do_apply, 3, 2}.
  {label,1}.
    {line,[{location,"src/callfunsmuggle.erl",5}]}.
    {func_info,{atom,callfunsmuggle},{atom,do_apply},3}.
  {label,2}.
    {'%',{var_info,{x,0},[{type,{t_fun,2,any,any}}]}}.
    {allocate,0,3}.
    {swap,{x,2},{x,1}}.
    {swap,{x,0},{x,2}}.
    {line,[{location,"src/callfunsmuggle.erl",6}]}.
    {call_fun2,{atom,safe},2,{tr,{x,2},{t_fun,2,any,any}}}.
    {deallocate,0}.
    return.

{function, send_msg, 2, 4}.
  {label,3}.
    {line,[{location,"src/callfunsmuggle.erl",8}]}.
    {func_info,{atom,callfunsmuggle},{atom,send_msg},2}.
  {label,4}.
    {move,{x,1},{x,2}}.
    {move,{x,0},{x,1}}.
    {move,{literal,fun erlang:send/2},{x,0}}.
    {call_only,3,{f,2}}. % do_apply/3
```

Though we might have to block `call_fun` and maybe `call_fun2` altogether: I think you could pass in an impure `Fun` through the environment of a transaction and `khepri_fun` wouldn't have a way to inspect it.

```erl
Self = self(),
Fun = fun erlang:send/2,
khepri:transaction(fun() ->
  Fun(Self, khepri_tx:get([a, b, c]))
end).
```

If it's a local call like `{call_fun2,{f,1},Arity,Tr}` we should be able to track it down though.

> I wonder if we need to patch the `{t_fun, ...}` tuple

It seems like `beam_validator` doesn't care about the arity. In fact, the instruction is accepted by `beam_validator` even with the `{Type, 0, 18446744073709551615}` part. `beam_validator` seems to be only checking the validity of the register holding the `Fun`.

> We should report this as a `beam_disasm` bug then?

I agree 👍

There's a decent sized set of mistranslations to report upstream [here](https://github.com/rabbitmq/khepri/blob/62197b4ac85351bef9501d17bb4b0a28c8b041f1/src/khepri_fun.erl#L942-L1047) that I think this one fits in with.

I think there might be a separate potential bug with `call_fun2` [here](https://github.com/erlang/otp/blob/694531bf5e893273fe7a47804aea4c4f4fef68d4/lib/compiler/src/beam_trim.erl#L253-L255) where the match on `I` will fail if `Fun` lives on a `y` register instead of an `x`. (Though it might be an intended reuse of the `I` binding: I can't find a way to force the `Fun` onto a y-reg, the compiler always `move/2`s it from back from a `y` to an `x` before the `call_fun2/3`).

**Comment by dumbbell** [comment]
> We should report this as a `beam_disasm` bug then?

Yes, at some point we will even submit several patches, based on the workarounds we have in `khepri_fun`.

**Comment by dumbbell** [comment]
> It looks like we'll need some extra checks for `move/2` to cover captures in order to track down `Fun`:

Variables passed through the `fun` environment are scanned and if there are `fun` in there, they are extracted:
1. https://github.com/rabbitmq/khepri/blob/01d55fe5e79c73fad3bbb0deba42884cdf7feb4d/src/khepri_fun.erl#L369
2. https://github.com/rabbitmq/khepri/blob/01d55fe5e79c73fad3bbb0deba42884cdf7feb4d/src/khepri_fun.erl#L1995-L2028

So hopefully, grounds are covered and we don't have to track arguments from the registers passed to `call_fun` and `call_fun2`.

**Comment by the-mikedavis** [comment]
Ah, makes sense!

I'll put together a PR with the `{atom,safe}` and `{atom,unsafe}` variants for `call_fun2/3` 👍
